### ISSUE TITLE
CI/CD feature request: add .drone.yml for drone-cli on local computer and drone CI server.

### ISSUE THREAD

**Issue by stevenleeS0ht** [sympy/sympy#21790]
Different developer have different local environment, CI workflow maybe difficult on local environment through GitHub workflow.

Add `.drone.yml` may be a good option for `drone-cli`.

Plus, in the future, `sympy` CI part may move to Drone service for fast execution speed.

`Drone` is a complete open source software, both its server and cli-tool are written in Golang, and just a single executable file.

Very easy to use.

**Comment by oscarbenjamin** [comment]
I don't know what drone is. What exactly are you suggesting?

**Comment by stevenleeS0ht** [comment]
@oscarbenjamin , 
Have a look?

https://www.drone.io/

https://github.com/drone/drone

https://github.com/drone/drone-cli

**Comment by oscarbenjamin** [comment]
I had a very quick look at each of those links and I don't understand why we should use this. Can you explain what you are suggesting and why?

**Comment by stevenleeS0ht** [comment]
@oscarbenjamin ,

At current moment, it is not for cloud usage, it is for local CI.

Github actions is too slow. for the majority of developer include me, we would rather execute those autotest in our own machine.

For some of us use `PyCharm`, we cannot install all the python tools for lint-check, coding style check in the repo's virtualenv, because it will make source code indexer so slow.

Drone is container native CI with parallelism enable. It would be ideal for local CI.

**Comment by sidhu1012** [comment]
@oscarbenjamin I think we should have one pep-8 check too,(something like pep-8-speaks) to check that the code follows PEP guidelines, and it shouldn't be mandatory to pass this test.

**Comment by asmeurer** [comment]
For the majority of PRs you can just run the tests with `./bin/test` or even just a subset of the tests (like `./bin/test solvers` or whatever). That will be much faster than GitHub actions because it won't be running multiple combinations of Python versions.

**Comment by moorepants** [comment]
> @oscarbenjamin I think we should have one pep-8 check too,(something like pep-8-speaks) to check that the code follows PEP guidelines, and it shouldn't be mandatory to pass this test.

@sidhu1012, the issue with implementing that in a library this big is that someone has to go through the whole library when the pep8 check is implemented and fix everything. You'll also find that there will be arguments on which pep8 rules should be strictly adhered to and loosely adhered to or ignored. I think the only way to do something like this for a library this big is adopt a tool like black and give up individual control on the style. black can be applied to the whole library in a minute or so. There are probably open issues about this and there are surely mailing list discussions.

**Comment by asmeurer** [comment]
We already have required flake8 tests, but it is configured to run only non-stylistic checks. If pep-8-speaks can be configured to us our flake8 configuration it might be useful to use it as it may give more informative messages than the CI failures.

**Comment by oscarbenjamin** [comment]
I'm not necessarily opposed to using something like black or a pep8 checker if it's not too onerous (I haven't tried using either before though so I'm not sure). Someone has to do the work though.

Having made the codebase flake8 compliant I can tell you that it was a lot of work. I think flake8 helps to pick up on real bugs though which is different from purely stylistic checking.

**Comment by asmeurer** [comment]
I'm opposed to black generally, and for SymPy specifically I think it tends to format things like mathematical expressions in a way that is worse than their current formatting (like spaces around all operators). 

Actually I think @sidhu1012's suggestion was to make the PEP 8 checks only apply to the new code for a given PR, and be non-mandatory. That could be useful (as long as it isn't too annoying), especially for new contributors who often have basic PEP 8 violations.

**Comment by oscarbenjamin** [comment]
The benchmark comments from the GitHub Actions bot could be extended to do this with diff-cover. It generates reports of coverage but can also report results from a quality tool:
https://diff-cover.readthedocs.io/en/latest/README.html

Using `pycodestyle` you can get either a text report like:
```
-------------
Diff Quality
Quality Report: pycodestyle
Diff: origin/master...HEAD, staged and unstaged changes
-------------
sympy/matrices/eigen.py (90.0%):
sympy/matrices/eigen.py:233: E231 missing whitespace after ','
-------------
Total:   10 lines
Violations: 1 line
% Quality: 90%
-------------
```
Or you can get a html report output but I'm not quite sure how to include that in a comment on github.

**Comment by sidhu1012** [comment]
> Actually I think @sidhu1012's suggestion was to make the PEP 8 checks only apply to the new code for a given PR, and be non-mandatory. That could be useful (as long as it isn't too annoying), especially for new contributors who often have basic PEP 8 violations. 

Yes precisely. I have seen it in many orgs  and it just reports the PEP8 violations with line number in the PR and also tells what the violation is. Ans we can also make pep-8-speaks bot to ignore certain pep8 guidelines as per our convenience.

**Comment by oscarbenjamin** [comment]
@sidhu1012 I'm not sure exactly what you are suggesting if you want to implement this then go ahead and then we can see how it looks.

**Comment by sidhu1012** [comment]
> @sidhu1012 I'm not sure exactly what you are suggesting if you want to implement this then go ahead and then we can see how it looks.

Yeah sure, I can try to implement this but for that @oscarbenjamin you may need to install/enable `pep-8-speaks` bot for repository as I can't do that since I don't have those rights.

**Comment by sidhu1012** [comment]
@oscarbenjamin You can check the bot [here](https://github.com/OrkoHunter/pep8speaks).

**Comment by oscarbenjamin** [comment]
You can enable that on your fork to test it out before we enable it on this repo.

**Comment by sidhu1012** [comment]
@oscarbenjamin If you want to see example you can check these two.
https://github.com/Dacience/DataScientist/pull/95
https://github.com/Dacience/DataScientist/pull/84#issuecomment-747871129

**Comment by sidhu1012** [comment]
> @oscarbenjamin If you want to see example you can check these two.
> https://github.com/Dacience/DataScientist/pull/84#issuecomment-747871129
> https://github.com/Dacience/DataScientist/pull/95#issuecomment-755441612

That's how the bot would comment and after every pushed commit would keep updating that comment. And it shouldn't be mandatory to pass this pep test.

**Comment by oscarbenjamin** [comment]
If you enable it on your fork then you can open a PR in your fork so we can see how it works and make sure we know what is needed to get it set up.
